src/runner_manager.py

```diff
@@ -0,0 +1,358 @@
+
+        with self.runner_bin_path.open(mode="wb") as runner_bin_file:
+            runner_bin_file.write(response.content)
+
+        # Verify the file integrity.
+        with tarfile.open(self.runner_bin_path, "r:gz") as f:
+            try:
+                logger.debug(
+                    "Downloaded GitHub runner binary contains files: %s", ", ".join(f.getnames())
+                )
+            except tarfile.TarError:
+                logger.error("Failed to decompress downloaded GitHub runner binary.")
+                raise
+
+    def get_github_info(self) -> List[RunnerInfo]:
+        """Get information on the runners from GitHub.
+
+        Returns:
+            List of information from GitHub on runners.
+        """
+        remote_runners = self._get_runner_github_info()
+        return [
+            RunnerInfo(r.name, r.status == "online", r.status == "offline")
+            for r in remote_runners.values()
+        ]
+
+    def reconcile(self, quantity: int, resources: VirtualMachineResources) -> int:
+        """Bring runners in line with target.
+
+        Args:
+            quantity: Number of intended runners.
+            resources: Configuration of the virtual machine resources.
+
+        Returns:
+            Difference between intended runners and actual runners.
+        """
+        runners = self._get_runners()
+
+        # Clean up offline runners
+        offline_runners = [r for r in runners if not r.online]
+        if offline_runners:
+            runner_names = ", ".join(r.name for r in offline_runners)
+            logger.info("Cleaning up offline runners: %s", runner_names)
+
+            for runner in offline_runners:
+                runner.remove()
+
+        # Add/Remove runners to match the target quantity
+        online_runners = [r for r in runners if r.exist and r.online]
+        delta = quantity - len(online_runners)
+        if delta > 0:
+            logger.info("Getting registration token for GitHub runners.")
+            if isinstance(self.path, GitHubRepo):
+                token = self._github.actions.create_registration_token_for_repo(
+                    owner=self.path.owner, repo=self.path.repo
+                )
+            else:
+                token = self._github.actions.create_registration_token_for_org(org=self.path.org)
+
+            logger.info("Adding %i additional runner(s)", delta)
+            for _ in range(delta):
+                runner = Runner(
+                    self._github,
+                    self._jinja,
+                    self._lxd,
+                    self.app_name,
+                    self.path,
+                    self.proxies,
+                    self._generate_runner_name(),
+                )
+                runner.create(
+                    self.image, resources, self.runner_bin_path, self.reconcile_interval, token
+                )
+        elif delta < 0:
+            idle_runners = [r for r in online_runners if not r.busy]
```

Comment:
In what case would the runners be busy and we would have too many runners?

Reply:
busy means GitHub has assigned a job to that runner.

Changes to the configuration on the number of runners while some existing runners are executing job can result in above scenario.

Reply:
Ah I see, makes sense